sourcemap/fleet_store.go

```diff
@@ -71,35 +79,74 @@ func newFleetStore(
 	fleetCfg *config.Fleet,
 	cfgs []config.SourceMapMetadata,
 ) (fleetStore, error) {
-	// TODO(stn): Add support for multiple fleet hosts
-	// cf. https://github.com/elastic/apm-server/issues/5514
-	host := fleetCfg.Hosts[0]
-	fleetURLs := make(map[key]string)
+	sourceMapURLs := make(map[key]string)
+	fleetBaseURLs := make([]string, len(fleetCfg.Hosts))
 
 	for _, cfg := range cfgs {
 		k := key{cfg.ServiceName, cfg.ServiceVersion, cfg.BundleFilepath}
-		u, err := common.MakeURL(fleetCfg.Protocol, cfg.SourceMapURL, host, 8220)
+		sourceMapURLs[k] = cfg.SourceMapURL
+	}
+
+	for i, host := range fleetCfg.Hosts {
+		baseURL, err := common.MakeURL(fleetCfg.Protocol, "", host, defaultFleetPort)
 		if err != nil {
 			return fleetStore{}, err
 		}
-		fleetURLs[k] = u
+		fleetBaseURLs[i] = baseURL
 	}
+
 	return fleetStore{
-		apikey:    "ApiKey " + fleetCfg.AccessAPIKey,
-		fleetURLs: fleetURLs,
-		c:         c,
+		apikey:        "ApiKey " + fleetCfg.AccessAPIKey,
+		fleetBaseURLs: fleetBaseURLs,
+		sourceMapURLs: sourceMapURLs,
+		c:             c,
+		rng:           rand.New(rand.NewSource(time.Now().UnixNano())),
 	}, nil
 }
 
 func (f fleetStore) fetch(ctx context.Context, name, version, path string) (string, error) {
 	k := key{name, version, path}
-	fleetURL, ok := f.fleetURLs[k]
+	sourceMapURL, ok := f.sourceMapURLs[k]
 	if !ok {
 		return "", fmt.Errorf("unable to find sourcemap.url for service.name=%s service.version=%s bundle.path=%s",
 			name, version, path,
 		)
 	}
 
+	idxs := f.rng.Perm(len(f.fleetBaseURLs))
+	respReceived := make(chan string, 1)
+	fetchErrReturned := make(chan error, 1)
```

Comment:
Does it make sense to limit the channel sizes to 1?

Reply:
In this particular case, I would say it's not necessary, and the way this code is written will actually create channels that can never be drained (and therefore also never GC'd, I think). The channels should also be closed when the function exits.

There are probably more use-cases, but I think I've only used buffered channels when I have a main "coordinating" go-routine that sends off work (eg., a main goroutine sending X messages to a smaller number of worker goroutines) and waits for a response/s.

Since the send on the channels is being done in each goroutine, and therefore not blocking any work, I don't think anything is gained by having a buffered channel, but @axw may have another opinion (or further bits of wisdom).